### PR TITLE
docs: Add Empower your application with Langchain, Langraph and Bedro…

### PR DESCRIPTION
**Pull request by viniciosneves-ustwo**

### CODE REVIEW

**Review by nickhegarty-ustwo — COMMENTED** [review]
A few suggestions on formal bits to address!

File: src/content/authors/vinicios-neves/index.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+name: Vinicios Neves
+role: Tech Lead
+contact: vinicios.neves@ustwo.com
+avatar: ./avatar-vinicios-neves.jpg
```
Comment (patriciagssilva) [inline src/content/authors/vinicios-neves/index.md:5]:
```suggestion
avatar: ./avatar-vinicios-neves.jpg
shortIntro: Vini's snow das neves.
contactInfo:
  - platform: Github
    url: <url>
```
Vini, now we also have shortIntro and contactInfo to add to the author section.

Available contact platforms are: **Email**, **Github**, **Instagram**, **LinkedIn**, and **X** (formerly Twitter).

Comment (patriciagssilva) [reply src/content/authors/vinicios-neves/index.md:5]: i.e: src/content/authors/patricia-silva/index.md

File: src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-1/index.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+So, if you’re ready to see how AI can level up your project—whether it’s for a simple chatbot or something more ambitious, like an app that suggests recipes based on what’s in your fridge—this article’s for you.
+
+Grab a coffee, let’s dive in!
+
+## Exploring the AI Landscape: Models, APIs, and Vendors
+
+When I first started diving into AI, one thing quickly became clear: there are a wide range of models out there, both open and closed-source. Then you have the challenge of integrating them through various APIs, each with its own quirks and limitations. Not to mention the vast ecosystem of packages designed to help you work with specific models or vendors.
+
+It was during this search that I stumbled upon **LangChain**. For someone like me, coming from a full stack development background, LangChain stood out as a way to simplify working with large language models (LLMs) without needing to reinvent the wheel.
```
[the remark this answers is not in the document]
Comment (viniciosneves-ustwo) [reply src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-1/index.md]: Done, links added:

- https://www.langchain.com/
- https://python.langchain.com/v0.2/docs/introduction/
- https://js.langchain.com/docs/introduction/

File: src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-1/index.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+For developers like us, who are always curious and eager to experiment with new tools, **LangChain** provides a way to easily tinker with models and rapidly prototype solutions. Whether you’re working with **Node.js** in a web-heavy project or sticking with **Python** for more research-oriented tasks, LangChain abstracts much of the complexity involved in interacting with LLMs. It enables you to integrate AI into your existing systems without having to reinvent the wheel, making it a valuable tool no matter which language you’re using.
+
+## Let’s Get Practical: A Simple Chat API with LangChain
+
+Let me show you an example of how easy it is to use LangChain to encapsulate a simple chat interaction, with no history or complex state to manage, using FastAPI in Python and Express in Node.js.
+
+### Python (FastAPI + LangChain)
+
+```py
+
```
[the remark this answers is not in the document]
Comment (viniciosneves-ustwo) [reply src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-1/index.md]: It means nothing, it was a glitch while exporting markdown from Google Docs (:

File: src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-1/index.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+});
+
+app.listen(3000, () => {
+  console.log('Server is running on port 3000');
+});
+```
+
+In the JavaScript example using Express, we achieve the same result by taking a prompt, sending it to OpenAI’s LLM, and returning the response—this time using LangChain.js. Here, we are specifically using the *text-davinci-003* model, which is a well-known **GPT-3** variant developed by OpenAI.
+
+*"text-davinci-003 is one of the most advanced GPT-3 models available in OpenAI’s suite of language models. It’s part of the GPT-3 family but is tuned to provide higher-quality and more detailed responses compared to earlier versions like text-curie or text-babbage. With its advanced capabilities, it excels at understanding complex instructions, handling nuanced tasks, and generating more coherent and contextually aware output."*
```
[the remark this answers is not in the document]
Comment (viniciosneves-ustwo) [reply src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-1/index.md]: It is not a blockquote, only an appendix explaining what "text-davinci-003" is, as Nayan asked.

File: src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-1/index.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+## Why LangChain?
```
[the remark this answers is not in the document]
Comment (viniciosneves-ustwo) [reply src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-1/index.md]: 💯 , done.

File: src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-1/index.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+What’s great about LangChain is how it simplifies the interaction with LLMs across different languages. Whether you’re a Python enthusiast or love working with JavaScript, LangChain handles all the complex communication with the LLM, leaving you to focus on building the rest of your app.
+
+If you’re a fullstack developer, you don’t need to go deep into the internals of AI models to start leveraging their power. LangChain wraps all of that for you, allowing you to integrate AI as just another tool in your development arsenal.
+
+That’s how I started adding AI to my projects. It wasn’t about becoming an AI expert, but about expanding my toolset as a fullstack developer. Whether it’s FastAPI or Express, LangChain makes adding AI simpler and more approachable.
+
```
[the remark this answers is not in the document]
Comment (viniciosneves-ustwo) [reply src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-1/index.md:100]: Added!

File: src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-1/index.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+author: Vinicios Neves
+date: 2024-10-09
+description: Learn how to integrate AI into your applications with LangChain, simplifying interactions with large language models (LLMs).
+thumbnail: ./assets/thumbnail.jpg
+tags: python, langchain, ai, javascript
+cta_prefix: AI is reshaping the future of app development. Join us as we explore the tools that make integrating AI easier for developers.
+---
+
+# Empower your application with Langchain, Langraph and Bedrock - Part 1: Exploring AI Tools
+
```
[the remark this answers is not in the document]
Comment (viniciosneves-ustwo) [reply src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-1/index.md:16]: Sure thing, added this to all of them:

## Table of Contents

1. [Empower your application with Langchain, Langraph and Bedrock - Part 1: Exploring AI Tools](/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-1)
2. [Empower your application with Langchain, Langraph and Bedrock - Part 2: Managing Complex AI Interactions with LangGraph](/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-2)
3. [Empower your application with Langchain, Langraph and Bedrock - Part 3: Leveraging Bedrock for Knowledge Base Management](/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-3)

File: src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-2/index.md [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+author: Vinicios Neves
+date: 2024-10-09
+description: Discover how LangGraph enables multi-step and multi-agent workflows for more sophisticated AI applications.
+thumbnail: ./assets/thumbnail.jpg
+tags: python, langchain, langgraph, ai
+cta_prefix: Take your AI interactions to the next level with LangGraph, a powerful tool for managing complex workflows in AI-driven apps.
+---
+
+# Empower your application with Langchain, Langraph and Bedrock - Part 2: Moving Beyond Simple Chats
+
```
[the remark this answers is not in the document]
Comment (viniciosneves-ustwo) [reply src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-2/index.md:16]: Added!

File: src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-2/index.md [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+**Step 3:** The AI may not be 100% sure, so it asks a clarifying question: “Are those fresh or dried herbs? Can you confirm?”  
+**Step 4:** Based on the answer, the AI refines its understanding and offers recipes that match the confirmed ingredients.
+
+The interaction becomes more sophisticated, where the agent doesn’t just respond, but also seeks clarification when needed. This shift moves us from simple question-answer exchanges to a more dynamic, **multi-modal** interaction—one where the agent can process different types of input and adapt its responses accordingly.
+
+For example, in a scenario where the AI is identifying ingredients from a photo of your fridge, it might not stop at merely describing what’s visible. If there’s ambiguity—like whether a green bunch is parsley or cilantro—the agent can ask follow-up questions or request more information. **O1** and similar models aim to push this even further by handling **multi-modal inputs**, meaning the agent isn’t restricted to text but can also interpret images, audio, and perhaps even video in real time.
```
[the remark this answers is not in the document]
Comment (viniciosneves-ustwo) [reply src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-2/index.md]: Yes, it should. Nice catch :)

File: src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-2/index.md [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+When building applications like the recipe suggestion app, you’re not just relying on one single model to handle everything. You’re dealing with a stack of components:
+
+* **Image recognition:** A model needs to process the photo, identify objects, and return a list of ingredients.  
+* **Questioning and clarification:** The agent might need to ask follow-up questions to refine its understanding.  
+* **Recipe generation:** Based on the confirmed ingredients, another model or system could be used to generate recipe suggestions.  
+* **State management:** Throughout all of this, the system needs to keep track of what it already knows, what it’s uncertain about, and how the user’s responses refine the final outcome.
+
+LangGraph helps with all of this by providing a structured way to manage these interactions. Instead of manually orchestrating every step, you can design a graph-based flow where different models and agents handle specific parts of the interaction.
+
+**Example: Recipe App with LangGraph**
```
[the remark this answers is not in the document]
Comment (viniciosneves-ustwo) [reply src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-2/index.md:58]: Not so easy, but added one! (:

File: src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-3/index.md [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+date: 2024-10-09
+description: Learn how to manage document-based knowledge bases with Amazon Bedrock, simplifying AI interactions for context-aware applications.
+thumbnail: ./assets/thumbnail.jpg
+tags: ai, bedrock, langchain, knowledge management
+cta_prefix: With tools like Bedrock, managing knowledge bases for AI-driven applications has never been easier.
+---
+
+# Empower your application with Langchain, Langraph and Bedrock \- Part 3: Leveraging Bedrock for Knowledge Base Management
+
+Welcome to the final part of our series\!
```
[the remark this answers is not in the document]
Comment (viniciosneves-ustwo) [reply src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-3/index.md]: Yep, all the weird markdown stuff was from GDocs exporting to the MD file.
Also, added the table of contents.

File: src/content/authors/vinicios-neves/index.md (hunk 1 above)
Comment (viniciosneves-ustwo) [reply src/content/authors/vinicios-neves/index.md:5]: Updated, thanks for the heads up!

**Review by patriciagssilva — APPROVED** [review]

File: src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-2/index.md [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+
+# Create Agents for Image Recognition, Clarification, and Recipe Generation
+def create_agent(llm, tools, system_message: str):
+    prompt = ChatPromptTemplate.from_messages(
+        [
+            (
+                "system",
+                "You are an AI assistant collaborating with other assistants. "
+                "Your task is to process the image, clarify ingredients, and generate a recipe."
+                " If your result is final, prefix it with 'FINAL ANSWER'."
```
Comment (nickhegarty-ustwo) [inline src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-2/index.md:129]: Note that your sample functions are appending "FINAL ANSWER", not prefixing with it - maybe we change this to just "If your result is final, include the string 'FINAL ANSWER' at the end"?

File: src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-2/index.md [hunk 13]
```diff
@@ -0,0 +1,306 @@
...
+
+At this point, we can manually control the retrieval of relevant documents based on user input through a retriever, which will search the knowledge base for the most relevant pieces of information. Then, using a system like Retrieval-Augmented Generation (RAG), the AI can generate responses based on those documents. Something like this:
+
+```py
+from fastapi import FastAPI
+from langchain.prompts import ChatPromptTemplate
+from langchain_openai import ChatOpenAI
+from dotenv import load_dotenv
+from langserve import add_routes
+import uvicorn
```
Comment (nickhegarty-ustwo) [inline src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-2/index.md:236]: Not a Python expert by any means, but should we group the import statements at the top, before the from statements, just for tidyness?

File: src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-1/index.md [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+date: 2024-10-09
+description: Learn how to integrate AI into your applications with LangChain, simplifying interactions with large language models (LLMs).
+thumbnail: ./assets/thumbnail.jpg
+tags: python, langchain, ai, javascript
+cta_prefix: AI is reshaping the future of app development. Join us as we explore the tools that make integrating AI easier for developers.
+---
+
+**Table of Contents**
+
+1. [Part 1: Exploring AI Tools](/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-1)
```
Comment (nickhegarty-ustwo) [inline src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-1/index.md:13]:
```suggestion
*This article is one of a multi-part series, which covers the use of Langchain, Langraph, and Bedrock for working with LLMs. Here are the other parts in the series:*

1. Part 1: Exploring AI Tools 👈🏼 **You are here**
```

**Review by nickhegarty-ustwo — COMMENTED** [review]
Looking great, a few more bits and a suggested change to the Table of Contents to be applied across all the articles if you're happy with it

**Review by nickhegarty-ustwo — COMMENTED** [review]
Minor changes to sample code

File: src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-3/index.md [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+description: Learn how to manage document-based knowledge bases with Amazon Bedrock, simplifying AI interactions for context-aware applications.
+thumbnail: ./assets/thumbnail.jpg
+tags: ai, bedrock, langchain, knowledge management
+cta_prefix: With tools like Bedrock, managing knowledge bases for AI-driven applications has never been easier.
+---
+
+*This article is one of a multi-part series, which covers the use of Langchain, Langraph, and Bedrock for working with LLMs. Here are the other parts in the series:*
+1. [Part 1: Exploring AI Tools](/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-1)
+2. Part 2: Managing Complex AI Interactions with LangGraph 👈🏼 **You are here**
+3. [Part 3: Leveraging Bedrock for Knowledge Base Management](/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-3)
```
Comment (nickhegarty-ustwo) [inline src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-3/index.md:14]:
```suggestion
2. [Part 2: Managing Complex AI Interactions with LangGraph](/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-2)
3. Part 3: Leveraging Bedrock for Knowledge Base Management 👈🏼 **You are here**
```

File: src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-3/index.md [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+tags: ai, bedrock, langchain, knowledge management
+cta_prefix: With tools like Bedrock, managing knowledge bases for AI-driven applications has never been easier.
+---
+
+*This article is one of a multi-part series, which covers the use of Langchain, Langraph, and Bedrock for working with LLMs. Here are the other parts in the series:*
+1. [Part 1: Exploring AI Tools](/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-1)
+2. [Part 2: Managing Complex AI Interactions with LangGraph](/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-2)
+3. Part 3: Leveraging Bedrock for Knowledge Base Management 👈🏼 **You are here**
+
+Welcome to the final part of our series!
```
Comment (nickhegarty-ustwo) [inline src/content/articles/empower-your-application-with-langchain-langraph-and-bedrock-part-3/index.md:16]:
```suggestion
Welcome to the final part of our series!

```

**Comment by nickhegarty-ustwo** [comment]
Merging Vini's blog posts 🎉

**Pull request merged**